Libs/vtkSegmentationCore/vtkBinaryLabelmapToClosedSurfaceConversionRule.cxx

```diff
@@ -288,24 +294,81 @@ bool vtkBinaryLabelmapToClosedSurfaceConversionRule::CreateClosedSurface(vtkOrie
   double smoothingFactor = this->ConversionParameters->GetValueAsDouble(GetSmoothingFactorParameterName());
   int computeSurfaceNormals = this->ConversionParameters->GetValueAsInt(GetComputeSurfaceNormalsParameterName());
 
-  vtkNew<vtkDiscreteFlyingEdges3D> marchingCubes;
-  marchingCubes->SetInputData(binaryLabelmapWithIdentityGeometry);
-  marchingCubes->ComputeGradientsOff();
-  marchingCubes->ComputeNormalsOff(); // While computing normals is faster using the flying edges filter,
-  // it results in incorrect normals in meshes from shared labelmaps marchingCubes->ComputeScalarsOn();
+  // Generation method
+  // 0 = flying edges
+  // 1 = surface nets
+  int generationMethod = this->ConversionParameters->GetValueAsInt(GetConversionMethodParameterName());
+  int surfaceNetsSmoothing = this->ConversionParameters->GetValueAsInt(GetInternalSmoothingParameterName());
+
+  std::cout << "Internal smoothing flag = " << surfaceNetsSmoothing << std::endl;
+
+  vtkSmartPointer<vtkPolyData> processingResult = vtkSmartPointer<vtkPolyData>::New();
+
+  if (generationMethod == 0)
+    {
+    vtkNew<vtkDiscreteFlyingEdges3D> flyingEdges;
+    flyingEdges->SetInputData(binaryLabelmapWithIdentityGeometry);
+    flyingEdges->ComputeGradientsOff();
+    flyingEdges->ComputeNormalsOff(); // While computing normals is faster using the flying edges filter,
+    // it results in incorrect normals in meshes from shared labelmaps marchingCubes->ComputeScalarsOn();
+
+    int valueIndex = 0;
+    for (vtkIdType labelValue : labelValues)
+      {
+      flyingEdges->SetValue(valueIndex, labelValue);
+      ++valueIndex;
+      }
+    try
+      {
+      flyingEdges->Update();
+      }
+    catch (...)
+      {
+      vtkErrorMacro("Convert: Error while running flying edges!");
+      return false;
+      }
+    processingResult = flyingEdges->GetOutput();
+    }
+  else if (generationMethod == 1)
+    {
+    vtkNew<vtkSurfaceNets3D> surfaceNets;
+    surfaceNets->SetInputData(binaryLabelmapWithIdentityGeometry);
+
+    // Disable internal smoothing, and use vtkWindowedSincPolyDataFilter for smoothing as needed
+    surfaceNets->SmoothingOff();
 
-  int valueIndex = 0;
-  for (vtkIdType labelValue : labelValues)
+    if (surfaceNetsSmoothing == 1)
     {
-    marchingCubes->SetValue(valueIndex, labelValue);
-    ++valueIndex;
+      surfaceNets->SmoothingOn();
+      int iterationCount = floor(10.0 * smoothingFactor);
```

Comment:
Add comment explaining the rational for choosing `10.0`

Reply:
I found a polynomial fit that behaves visually similarly. Check it out once I do another push.

Reply:
For future reference:

```
      // This formula maps (input) -> (iteration count)
      // 0.0  ->  0   (almost no smoothing)
      // 0.2  ->  2   (little smoothing)
      // 0.5  ->  8   (average smoothing)
      // 0.7  ->  14  (strong smoothing)
      // 1.0  ->  24  (very strong smoothing)
```